server/src/main/java/org/elasticsearch/bootstrap/Spawner.java

```diff
@@ -34,22 +40,28 @@ final class Spawner implements Closeable {
      * References to the processes that have been spawned, so that we can destroy them.
      */
     private final List<Process> processes = new ArrayList<>();
+    private final List<Thread> pumpThreads = new ArrayList<>();
     private AtomicBoolean spawned = new AtomicBoolean();
 
     @Override
     public void close() throws IOException {
-        IOUtils.close(() -> processes.stream().map(s -> (Closeable) s::destroy).iterator());
+        List<Closeable> closeables = new ArrayList<>();
+        closeables.addAll(processes.stream().map(s -> (Closeable) s::destroy).toList());
+        closeables.addAll(pumpThreads.stream().map(t -> (Closeable) () -> {
+            try {
+                t.join(); // wait for thread to complete now that the spawned process is destroyed
+            } catch (InterruptedException e) { /* best effort, ignore*/ }
```

Comment:
I think it would be best practice to call `Thread.currentThread().interrupt();` in the `catch` block here, because if the interruption occurred when joining the first thread we wouldn't want the join of the second thread to block.

Reply:
Fair enough, I pushed [f68b57f](https://github.com/elastic/elasticsearch/pull/86709/commits/f68b57faa341dc622eaafcd30fd64e602499ceeb)